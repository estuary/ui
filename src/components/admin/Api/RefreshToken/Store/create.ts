import produce from 'immer';
import { BASE_ERROR } from 'services/supabase';
import { devtoolsOptions } from 'utils/store-utils';
import type { StoreApi } from 'zustand';
import { create } from 'zustand';
import type { NamedSet } from 'zustand/middleware';
import { devtools } from 'zustand/middleware';
import type { RefreshTokenState } from './types';

const getInitialStateData = (): Pick<
    RefreshTokenState,
    'description' | 'saving' | 'serverError' | 'token'
> => ({
    description: '',
    saving: false,
    serverError: null,
    token: '',
});

const getInitialState = (
    set: NamedSet<RefreshTokenState>,
    _get: StoreApi<RefreshTokenState>['getState']
): RefreshTokenState => ({
    ...getInitialStateData(),

    resetState: () => {
        set(getInitialStateData(), false, 'State reset');
    },

    setSaving: (value) => {
        set(
            produce((state: RefreshTokenState) => {
                state.saving = value;
            }),
            false,
            'Saving set'
        );
    },

    setServerError: (value) => {
        set(
            produce((state: RefreshTokenState) => {
                state.serverError =
                    typeof value === 'string'
                        ? { ...BASE_ERROR, message: value }
                        : value;
            }),
            false,
            'Server error set'
        );
    },

    setToken: (value) => {
        set(
            produce((state: RefreshTokenState) => {
                state.token = value;
            }),
            false,
            'Refresh token set'
        );
    },

    updateDescription: (value) => {
        set(
            produce((state: RefreshTokenState) => {
                state.description = value;
            }),
            false,
            'Token description updated'
        );
    },
});

export const useRefreshTokenStore = create<RefreshTokenState>()(
    devtools(
        (set, get) => getInitialState(set, get),
        devtoolsOptions('refresh-tokens')
    )
);
