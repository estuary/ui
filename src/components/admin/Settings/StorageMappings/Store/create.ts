import produce from 'immer';
import { devtoolsOptions } from 'utils/store-utils';
import { StoreApi, create } from 'zustand';
import { NamedSet, devtools } from 'zustand/middleware';
import { StorageMappingState } from './types';

const getInitialStateData = (): Pick<
    StorageMappingState,
    'formValue' | 'logToken' | 'provider' | 'pubId' | 'saving' | 'serverError'
> => ({
    formValue: { data: {} },
    logToken: '',
    provider: '',
    pubId: '',
    saving: false,
    serverError: null,
});

const getInitialState = (
    set: NamedSet<StorageMappingState>,
    _get: StoreApi<StorageMappingState>['getState']
): StorageMappingState => ({
    ...getInitialStateData(),

    resetState: () => {
        set(getInitialStateData(), false, 'State reset');
    },

    setSaving: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.saving = value;
            }),
            false,
            'Saving set'
        );
    },

    setServerError: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.serverError =
                    typeof value === 'string' ? { message: value } : value;
            }),
            false,
            'Server error set'
        );
    },

    setLogToken: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.logToken = value;
            }),
            false,
            'Publication log token set'
        );
    },

    setPubId: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.pubId = value;
            }),
            false,
            'Publication ID set'
        );
    },

    updateFormValue: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.formValue = value;
            }),
            false,
            'Form updated'
        );
    },

    updateProvider: (value) => {
        set(
            produce((state: StorageMappingState) => {
                state.formValue = { data: {} };
                state.provider = value;
            }),
            false,
            'Cloud provider updated'
        );
    },
});

export const useStorageMappingStore = create<StorageMappingState>()(
    devtools(
        (set, get) => getInitialState(set, get),
        devtoolsOptions('storage-mappings')
    )
);
