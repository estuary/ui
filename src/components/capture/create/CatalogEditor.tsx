import Editor from '@monaco-editor/react';
import {
    Box,
    List,
    ListItemButton,
    ListItemText,
    Paper,
    Typography,
    useTheme,
} from '@mui/material';
import * as monacoEditor from 'monaco-editor/esm/vs/editor/editor.api';
import { useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { TABLES } from 'services/supabase';
import useSchemaEditorStore, {
    SchemaEditorState,
} from 'stores/SchemaEditorStore';
import { useQuery, useSelect } from 'supabase-swr';

type Props = {
    draftId: string;
};

interface DraftSpec {
    catalog_name: string;
    spec_type: string;
    spec_patch: string;
    draft_id: string;
}

const CONNECTOR_TAGS_QUERY = `
    catalog_name,
    spec_type,
    spec_patch,
    draft_id
`;

const selectors = {
    loadResource: (state: SchemaEditorState) => state.loadResource,
    updateResource: (state: SchemaEditorState) => state.updateResource,
};

function NewCaptureEditor({ draftId }: Props) {
    const loadResource = useSchemaEditorStore(selectors.loadResource);
    const updateResource = useSchemaEditorStore(selectors.updateResource);

    const draftSpecQuery = useQuery<DraftSpec>(
        TABLES.DRAFT_SPECS,
        {
            columns: CONNECTOR_TAGS_QUERY,
            filter: (query) => query.eq('draft_id', draftId),
        },
        []
    );
    const { data } = useSelect(draftSpecQuery, {});
    const draftSpecs: DraftSpec[] = data ? data.data : [];

    const theme = useTheme();

    const editorRef = useRef<monacoEditor.editor.IStandaloneCodeEditor | null>(
        null
    );

    const [currentFileIndex, setCurrentFileIndex] = useState(0);

    const handlers = {
        fileList: {
            click: (tagInex: number) => {
                setCurrentFileIndex(tagInex);
            },
        },
        change: () => {
            if (editorRef.current) {
                updateResource(
                    draftSpecs[currentFileIndex].catalog_name,
                    editorRef.current.getValue()
                );
            }
        },
        mount: (editor: monacoEditor.editor.IStandaloneCodeEditor) => {
            editorRef.current = editor;

            handlers.fileList.click(0);

            if (draftSpecs.length > 0) {
                draftSpecs.forEach((tag) => {
                    loadResource(tag.catalog_name, tag.spec_patch);
                });
            }

            // Commented out as it stands as the main example of how to handle "events"
            // const handler = editor.onDidChangeModelDecorations(() => {
            //     handler.dispose();
            //     void editor.getAction('editor.action.formatDocument').run();
            // });
        },
    };

    return (
        <>
            <Typography variant="h5">Catalog Editor</Typography>
            <Typography>
                <FormattedMessage id="captureCreation.finalReview.instructions" />
            </Typography>
            <Paper variant="outlined">
                {draftSpecs.length > 0 ? (
                    <Box
                        sx={{
                            flexGrow: 1,
                            bgcolor: 'background.paper',
                            display: 'flex',
                            height: 300,
                        }}
                    >
                        <List dense disablePadding>
                            {draftSpecs.map((tag: any, index: number) => (
                                <ListItemButton
                                    key={`FileSelector-${tag.catalog_name}`}
                                    dense
                                    selected={index === currentFileIndex}
                                    onClick={() => {
                                        handlers.fileList.click(index);
                                    }}
                                >
                                    <ListItemText
                                        primary={tag.catalog_name}
                                        secondary={tag.spec_type}
                                    />
                                </ListItemButton>
                            ))}
                        </List>
                        <Editor
                            height="300px"
                            defaultLanguage="json"
                            theme={
                                theme.palette.mode === 'light'
                                    ? 'vs'
                                    : 'vs-dark'
                            }
                            defaultValue={JSON.stringify(
                                draftSpecs[currentFileIndex],
                                null,
                                2
                            )}
                            path={draftSpecs[currentFileIndex].catalog_name}
                            onMount={handlers.mount}
                            onChange={handlers.change}
                        />
                    </Box>
                ) : (
                    <FormattedMessage id="common.loading" />
                )}
            </Paper>
        </>
    );
}

export default NewCaptureEditor;
