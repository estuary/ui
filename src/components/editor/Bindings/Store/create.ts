import { extend_read_bundle, infer } from '@estuary/flow-web';
import {
    createDraftSpec,
    getDraftSpecsByCatalogName,
    modifyDraftSpec,
} from 'api/draftSpecs';
import { fetchInferredSchema } from 'api/inferred_schemas';
import { getLiveSpecsByCatalogName } from 'api/liveSpecsExt';
import { BindingsEditorState } from 'components/editor/Bindings/Store/types';
import { CollectionData } from 'components/editor/Bindings/types';
import produce from 'immer';
import { forEach, intersection, isEmpty, isPlainObject, union } from 'lodash';
import { Dispatch, SetStateAction } from 'react';
import { logRocketEvent } from 'services/shared';
import { CallSupabaseResponse } from 'services/supabase';
import { BindingsEditorStoreNames } from 'stores/names';
import { checkForErrors } from 'stores/utils';
import {
    InferSchemaPropertyForRender,
    InferSchemaResponse,
    Schema,
} from 'types';
import { hasLength } from 'utils/misc-utils';
import { filterInferSchemaResponse, hasReadSchema } from 'utils/schema-utils';
import { devtoolsOptions } from 'utils/store-utils';
import { getSourceOrTarget } from 'utils/workflow-utils';
import { StoreApi, create } from 'zustand';
import { NamedSet, devtools } from 'zustand/middleware';

const processDraftSpecResponse = (
    draftSpecResponse: CallSupabaseResponse<any>,
    setOpen: Dispatch<SetStateAction<boolean>>,
    get: StoreApi<BindingsEditorState>['getState']
) => {
    const {
        inferredSchemaApplicationErrored,
        setCollectionData,
        setInferredSchemaApplicationErrored,
    } = get();

    if (draftSpecResponse.error) {
        setInferredSchemaApplicationErrored(true);
    } else if (draftSpecResponse.data && draftSpecResponse.data.length > 0) {
        if (inferredSchemaApplicationErrored) {
            setInferredSchemaApplicationErrored(false);
        }

        setCollectionData({
            spec: draftSpecResponse.data[0].spec,
            belongsToDraft: true,
        });

        setOpen(false);
    } else {
        setInferredSchemaApplicationErrored(true);
    }
};

const evaluateCollectionData = async (
    draftId: string | null,
    catalogName: string
): Promise<CollectionData | null> => {
    let draftSpecResponse = null;

    if (draftId) {
        draftSpecResponse = await getDraftSpecsByCatalogName(
            draftId,
            catalogName,
            'collection'
        );
    }

    if (draftSpecResponse?.data && !isEmpty(draftSpecResponse.data)) {
        return { spec: draftSpecResponse.data[0].spec, belongsToDraft: true };
    } else {
        const liveSpecResponse = await getLiveSpecsByCatalogName(
            catalogName,
            'collection'
        );

        return !liveSpecResponse.data || isEmpty(liveSpecResponse.data)
            ? null
            : { spec: liveSpecResponse.data[0].spec, belongsToDraft: false };
    }
};

// Used to properly populate the inferSchemaResponse related state
const evaluateInferSchemaResponse = (dataVal: InferSchemaResponse[] | null) => {
    let updatedVal: InferSchemaPropertyForRender[] | null, validKeys: string[];

    const hasResponse = dataVal && dataVal.length > 0;
    if (hasResponse) {
        // Need list of fields and keys to be populated by the data
        const filteredKeys: string[][] = [];
        const filteredFields: any[] = [];

        // Go through all the data and grab the filtered values
        //  and put them into the lists
        forEach(dataVal, (val) => {
            const response = filterInferSchemaResponse(val);
            filteredKeys.push(response.validKeys);
            filteredFields.push(response.fields);
        });

        // Find the keys that are valid in BOTH read/write schema
        validKeys = intersection(...filteredKeys);

        // Put the read/write output together so all keys are rendered
        updatedVal = union(...filteredFields);
    } else {
        updatedVal = null;
        validKeys = [];
    }

    return {
        updatedVal,
        validKeys,
        hasResponse,
    };
};

// Call into the flow WASM handler that will inline the write/inferred schema if necessary
const updateReadSchema = async (
    read: Schema,
    write: Schema,
    entityName: string
) => {
    // Try fetching the inferred schema... possible TODO handle errors better
    const inferredSchemaResponse = await fetchInferredSchema(entityName);
    const inferred = inferredSchemaResponse.data?.[0]?.schema
        ? inferredSchemaResponse.data[0].schema
        : {};

    let response;
    try {
        response = extend_read_bundle({
            read,
            write,
            inferred,
        });
        // We can catch any error here so that any issue causes an empty response and the
        //  component will show an error... though not the most useful one.
        // eslint-disable-next-line @typescript-eslint/no-implicit-any-catch
    } catch (e: any) {
        logRocketEvent('extend_read_bundle:failed', e);
        response = {};
    }
    return response;
};

const getInitialFullSourceData = (): Pick<
    BindingsEditorState,
    'fullSourceConfigs' | 'fullSourceErrorsExist'
> => ({
    fullSourceConfigs: {},
    fullSourceErrorsExist: false,
});

const getInitialFieldSelectionData = (): Pick<
    BindingsEditorState,
    'selections' | 'selectionSaving'
> => ({
    selections: {},
    selectionSaving: false,
});

const getInitialMiscData = (): Pick<
    BindingsEditorState,
    | 'collectionData'
    | 'collectionInitializationAlert'
    | 'documentsRead'
    | 'inferredSchemaApplicationErrored'
    | 'inferredSpec'
    | 'loadingInferredSchema'
    | 'schemaInferenceDisabled'
    | 'schemaUpdateErrored'
    | 'schemaUpdated'
    | 'editModeEnabled'
    | 'inferSchemaResponse'
    | 'inferSchemaResponseError'
    | 'inferSchemaResponseDoneProcessing'
    | 'inferSchemaResponseEmpty'
    | 'inferSchemaResponse_Keys'
    | 'incompatibleCollections'
    | 'hasIncompatibleCollections'
    | 'selections'
    | 'selectionSaving'
> => ({
    collectionData: null,
    collectionInitializationAlert: null,
    documentsRead: null,
    inferredSchemaApplicationErrored: false,
    inferredSpec: null,
    loadingInferredSchema: false,
    schemaInferenceDisabled: false,
    schemaUpdateErrored: false,
    schemaUpdated: true,
    editModeEnabled: false,
    inferSchemaResponse: null,
    inferSchemaResponse_Keys: [],
    inferSchemaResponseError: null,
    inferSchemaResponseDoneProcessing: false,
    inferSchemaResponseEmpty: false,
    incompatibleCollections: [],
    hasIncompatibleCollections: false,
    ...getInitialFieldSelectionData(),
});

const getInitialStateData = () => ({
    ...getInitialMiscData(),
    ...getInitialFullSourceData(),
});

const getInitialState = (
    set: NamedSet<BindingsEditorState>,
    get: StoreApi<BindingsEditorState>['getState']
): BindingsEditorState => ({
    ...getInitialStateData(),

    setEditModeEnabled: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.editModeEnabled = value;
            }),
            false,
            'Edit Mode Enabled Set'
        );
    },

    setCollectionData: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.collectionData = value;
            }),
            false,
            'Collection Data Set'
        );
    },

    setCollectionInitializationAlert: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.collectionInitializationAlert = value;
            }),
            false,
            'Collection Initialization Alert Set'
        );
    },

    setIncompatibleCollections: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.incompatibleCollections = value;
                state.hasIncompatibleCollections = hasLength(value);
            }),
            false,
            'Incompatible Collections List Set'
        );
    },

    setSchemaInferenceDisabled: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.schemaInferenceDisabled = value;
            }),
            false,
            'Schema Inference Disabled Set'
        );
    },

    setLoadingInferredSchema: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.loadingInferredSchema = value;
            }),
            false,
            'Loading Inferred Schema Set'
        );
    },

    setInferredSpec: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.inferredSpec = value;
            }),
            false,
            'Inferred Schema Set'
        );
    },

    setDocumentsRead: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.documentsRead = value;
            }),
            false,
            'Documents Read Set'
        );
    },

    setInferredSchemaApplicationErrored: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.inferredSchemaApplicationErrored = value;
            }),
            false,
            'Inferred Schema Application Errored Set'
        );
    },

    removeFullSourceConfig: (collection) => {
        set(
            produce((state: BindingsEditorState) => {
                // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
                delete state.fullSourceConfigs[collection];
            }),
            false,
            'Removing full source config of a collection'
        );
    },

    updateFullSourceConfig: (collection, formData) => {
        set(
            produce((state: BindingsEditorState) => {
                const existingData =
                    state.fullSourceConfigs[collection]?.data ?? {};
                const fullSource = formData.data;

                state.fullSourceConfigs[collection] = {
                    data: {
                        ...existingData,
                        ...fullSource,
                    },
                    errors: formData.errors,
                };

                state.fullSourceErrorsExist = checkForErrors(formData)
                    ? true
                    : Object.values(state.fullSourceConfigs).some(
                          (fullSourceConfig) => checkForErrors(fullSourceConfig)
                      );
            }),
            false,
            'Updating full source config of a collection'
        );
    },

    prefillFullSourceConfigs: (bindings) => {
        set(
            produce((state: BindingsEditorState) => {
                const newConfig = {};

                if (bindings && bindings.length > 0) {
                    bindings.forEach((binding) => {
                        const bindingSource = getSourceOrTarget(binding);
                        const nameOnly = typeof bindingSource === 'string';

                        if (nameOnly) {
                            newConfig[bindingSource] = {};
                        } else {
                            const { name, ...restOfFullSource } = bindingSource;
                            newConfig[name] = {
                                data: restOfFullSource,
                                errors: [],
                            };
                        }
                    });
                }

                state.fullSourceConfigs = newConfig;
            }),
            false,
            'Prefilling full source configs'
        );
    },

    applyInferredSchema: async (
        currentCollection,
        persistedDraftId,
        setOpen
    ) => {
        const {
            collectionData,
            inferredSpec,
            setInferredSchemaApplicationErrored,
            setLoadingInferredSchema,
        } = get();

        if (
            currentCollection &&
            persistedDraftId &&
            inferredSpec &&
            collectionData
        ) {
            setInferredSchemaApplicationErrored(false);
            setLoadingInferredSchema(true);

            if (collectionData.belongsToDraft) {
                const draftSpecResponse = await modifyDraftSpec(inferredSpec, {
                    draft_id: persistedDraftId,
                    catalog_name: currentCollection,
                });

                processDraftSpecResponse(draftSpecResponse, setOpen, get);
            } else {
                const liveSpecsResponse = await getLiveSpecsByCatalogName(
                    currentCollection,
                    'collection'
                );

                if (liveSpecsResponse.error) {
                    setInferredSchemaApplicationErrored(true);
                } else if (liveSpecsResponse.data) {
                    const { last_pub_id } = liveSpecsResponse.data[0];

                    const draftSpecResponse = await createDraftSpec(
                        persistedDraftId,
                        currentCollection,
                        inferredSpec,
                        'collection',
                        last_pub_id
                    );

                    processDraftSpecResponse(draftSpecResponse, setOpen, get);
                }
            }

            setLoadingInferredSchema(false);
        } else {
            setInferredSchemaApplicationErrored(true);
        }
    },

    setSchemaUpdateErrored: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.schemaUpdateErrored = value;
            }),
            false,
            'Schema Update Errored Set'
        );
    },

    setSchemaUpdated: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.schemaUpdated = value;
            }),
            false,
            'Schema Updated Set'
        );
    },

    // TODO (optimization): Equip stores with the proper tools to clean up after themselves; this includes managing the promises they create.
    updateSchema: (currentCollection, persistedDraftId) => {
        const {
            collectionData,
            schemaUpdateErrored,
            setSchemaUpdateErrored,
            setSchemaUpdated,
            setCollectionData,
        } = get();

        if (currentCollection && collectionData) {
            setSchemaUpdated(false);

            return evaluateCollectionData(
                persistedDraftId,
                currentCollection
            ).then(
                (response) => {
                    if (schemaUpdateErrored) {
                        setSchemaUpdateErrored(false);
                    }

                    setCollectionData(response);

                    setSchemaUpdated(true);
                },
                () => {
                    setSchemaUpdateErrored(true);
                    setSchemaUpdated(true);
                }
            );
        } else {
            return null;
        }
    },

    setSingleSelection: (field, selectionType) => {
        set(
            produce((state: BindingsEditorState) => {
                const { selectionSaving } = get();

                state.selections[field] = selectionType;

                if (!selectionSaving) {
                    state.selectionSaving = true;
                }
            }),
            false,
            'Custom Selections Set'
        );
    },

    setSelectionSaving: (value) => {
        set(
            produce((state: BindingsEditorState) => {
                state.selectionSaving = value;
            }),
            false,
            'Selection Saving Set'
        );
    },

    // TODO (collection editor) maybe
    // This code was initially written supporting being able to
    //  run against `schema` OR [`readSchema`, `writeSchema`].
    // That was removed but might be needed in the future
    //  so we left things running through loops in case we need
    //  to support that again
    populateInferSchemaResponse: async (spec, entityName) => {
        const populateState = (
            dataVal: InferSchemaResponse[] | null,
            errorVal: BindingsEditorState['inferSchemaResponseError']
        ) => {
            const { hasResponse, updatedVal, validKeys } =
                evaluateInferSchemaResponse(dataVal);

            // Save the values into the store
            set(
                produce((state: BindingsEditorState) => {
                    state.inferSchemaResponseError = errorVal;
                    state.inferSchemaResponse = updatedVal;
                    state.inferSchemaResponseEmpty = !hasResponse;
                    state.inferSchemaResponse_Keys = validKeys;
                    state.inferSchemaResponseDoneProcessing = true;
                }),
                false,
                'Infere Schema Populated'
            );
        };

        set(
            produce((state: BindingsEditorState) => {
                state.inferSchemaResponseDoneProcessing = false;
            }),
            false,
            'Resetting inferSchemaDoneProcessing flag'
        );

        // If no schema then just return because hopefully it means
        //  we are still just waiting for the schema to load in
        if (!spec) {
            return;
        }

        // Check which schema to use
        const usingReadSchema = hasReadSchema(spec);
        const schemasToTest = usingReadSchema
            ? [spec.readSchema]
            : [spec.schema];

        // If no schema then just return because hopefully it means
        //  we are still just waiting for the schema to load in
        if (!hasLength(schemasToTest)) {
            return;
        }

        // Make sure we have an object
        if (!schemasToTest.every((schema) => isPlainObject(schema))) {
            populateState(null, 'schema must be an object');
            return;
        }

        try {
            // Should only impact the read schema
            if (usingReadSchema) {
                // We MUST make this call before calling `infer` below
                //  This will inline the write/inferred schema in the `$defs` if needed
                schemasToTest[0] = await updateReadSchema(
                    schemasToTest[0],
                    spec.writeSchema ?? {},
                    entityName
                );
            }

            // Run infer against schema
            const responses = schemasToTest.map((schema) => infer(schema));

            // Make sure all the responses are valid
            const allResponsesValid = responses.every((inferResponse) => {
                const { properties } = inferResponse;

                // Make sure there is a response
                if (properties?.length === 0) {
                    populateState(null, 'no fields inferred from schema');
                    return false;
                }

                // Make sure we did not ONLY get the root object back as a pointer
                if (properties.length === 1 && properties[0].pointer === '') {
                    populateState(
                        null,
                        'no usable fields inferred from schema'
                    );
                    return false;
                }

                return true;
            });

            if (!allResponsesValid) {
                // This just returns because we already populated the state
                //  in the above loop
                return;
            }

            populateState(responses, null);
        } catch (err: unknown) {
            populateState(null, err as string);
        }
    },

    resetState: (skipFullSource) => {
        set(
            skipFullSource ? getInitialMiscData() : getInitialStateData(),
            false,
            'Bindings Editor State Reset'
        );
    },
});

export const useBindingsEditorStore = create<BindingsEditorState>()(
    devtools(
        (set, get) => getInitialState(set, get),
        devtoolsOptions(BindingsEditorStoreNames.GENERAL)
    )
);
