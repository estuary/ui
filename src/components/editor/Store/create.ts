import type { EditorStoreState } from 'src/components/editor/Store/types';
import type { NamedSet } from 'zustand/middleware';

import { create } from 'zustand';
import { devtools } from 'zustand/middleware';

import produce from 'immer';
import { omit } from 'lodash';

import { EditorStatus } from 'src/components/editor/Store/types';
import { devtoolsOptions } from 'src/utils/store-utils';

export const isEditorActive = (status: EditorStatus) => {
    return status === EditorStatus.SAVING;
};

const getInitialStateData = <T>(): Pick<
    EditorStoreState<T>,
    | 'catalogName'
    | 'currentCatalog'
    | 'discoveredDraftId'
    | 'draftInitializationError'
    | 'id'
    | 'invalidEditors'
    | 'isEditing'
    | 'isSaving'
    | 'persistedDraftId'
    | 'pubId'
    | 'queryResponse'
    | 'serverUpdate'
    | 'specs'
    | 'statuses'
> => {
    return {
        catalogName: '',
        currentCatalog: null,
        id: null,
        discoveredDraftId: null,
        persistedDraftId: null,
        pubId: null,
        specs: null,
        statuses: {},
        isSaving: false,
        isEditing: false,
        serverUpdate: null,
        draftInitializationError: null,
        queryResponse: { draftSpecs: [], isValidating: false, mutate: null },
        invalidEditors: [],
    };
};

const getInitialState = <T>(
    set: NamedSet<EditorStoreState<T>>
): EditorStoreState<T> => {
    return {
        ...getInitialStateData<T>(),
        setId: (newVal) => {
            set(
                produce((state) => {
                    state.id = newVal;
                }),
                false,
                'Set draft id'
            );
        },

        setCatalogName: (newVal) => {
            set(
                produce((state) => {
                    state.catalogName = newVal;
                }),
                false,
                'Set catalog name'
            );
        },

        setPersistedDraftId: (newVal) => {
            set(
                produce((state) => {
                    state.persistedDraftId = newVal;
                }),
                false,
                'Set persisted draft id'
            );
        },

        setDiscoveredDraftId: (newVal) => {
            set(
                produce((state) => {
                    state.discoveredDraftId = newVal;
                }),
                false,
                'Set discovered draft id'
            );
        },

        setPubId: (newVal) => {
            set(
                produce((state) => {
                    state.pubId = newVal;
                }),
                false,
                'Set publication id'
            );
        },

        setCurrentCatalog: (newVal) => {
            set(
                produce((state) => {
                    state.currentCatalog = newVal;
                    state.status = EditorStatus.IDLE;
                }),
                false,
                'Setting current catalog'
            );
        },

        setSpecs: (newVal) => {
            set(
                produce((state) => {
                    if (newVal && newVal.length > 0) {
                        if (state.specs === null || newVal.length === 1) {
                            state.currentCatalog = newVal[0];
                        }
                        state.specs = newVal;
                    }
                }),
                false,
                'Set specs'
            );
        },

        setServerUpdate: (newVal) => {
            set(
                produce((state) => {
                    state.serverUpdate = newVal;
                }),
                false,
                'Set server update'
            );
        },

        setStatus: (value, path) => {
            set(
                produce((state) => {
                    state.statuses[path] = value;

                    const editorStatuses: EditorStatus[] = Object.values(
                        state.statuses
                    );

                    state.isSaving = editorStatuses.some(
                        (status) => status === EditorStatus.SAVING
                    );
                    state.isEditing = editorStatuses.some(
                        (status) => status === EditorStatus.EDITING
                    );

                    state.invalidEditors = Object.entries(state.statuses)
                        .filter(
                            ([_key, status]) =>
                                status === EditorStatus.INVALID ||
                                status === EditorStatus.SAVE_FAILED ||
                                status === EditorStatus.OUT_OF_SYNC
                        )
                        .map(([key, _status]) => key);
                }),
                false,
                'Setting status'
            );
        },

        removeStaleStatus: (value) => {
            set(
                produce((state: EditorStoreState<any>) => {
                    if (Object.hasOwn(state.statuses, value)) {
                        state.statuses = omit(state.statuses, value);
                    }

                    state.invalidEditors = state.invalidEditors.filter(
                        (path) => path !== value
                    );
                }),
                false,
                'Stale Status Removed'
            );
        },

        setDraftInitializationError: (value) => {
            set(
                produce((state) => {
                    state.draftInitializationError = value;
                }),
                false,
                'Draft Initialization Error Set'
            );
        },

        setQueryResponse: (value) => {
            set(
                produce((state) => {
                    state.queryResponse = value;
                }),
                false,
                'Query Response Set'
            );
        },

        // This is a hacky, temporary solution to preserve the persisted draft ID
        // when the generate button is clicked in all workflows.
        resetState: (excludePersistedDraftId) => {
            set(
                () => {
                    const { persistedDraftId, ...rest } =
                        getInitialStateData<T>();

                    return excludePersistedDraftId
                        ? rest
                        : { persistedDraftId, ...rest };
                },
                false,
                'Resetting Editor State'
            );
        },
    };
};

export const createEditorStore = <T>(key: string) => {
    return create<EditorStoreState<T>>()(
        devtools((set) => getInitialState<T>(set), devtoolsOptions(key))
    );
};
