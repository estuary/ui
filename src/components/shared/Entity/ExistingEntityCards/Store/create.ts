import { getLiveSpecsByConnectorId } from 'api/liveSpecsExt';
import { ExistingEntityState } from 'components/shared/Entity/ExistingEntityCards/Store/types';
import produce from 'immer';
import {
    getInitialHydrationData,
    getStoreWithHydrationSettings,
} from 'stores/extensions/Hydration';
import { ExistingEntityStoreNames } from 'stores/names';
import { devtoolsOptions } from 'utils/store-utils';
import { create, StoreApi } from 'zustand';
import { devtools, NamedSet } from 'zustand/middleware';

const getInitialStateData = (): Pick<
    ExistingEntityState,
    'connectorName' | 'createNewTask' | 'queryData'
> => ({
    connectorName: null,
    createNewTask: false,
    queryData: null,
});

const getInitialState = (
    set: NamedSet<ExistingEntityState>,
    get: StoreApi<ExistingEntityState>['getState']
): ExistingEntityState => {
    return {
        ...getInitialStateData(),
        ...getStoreWithHydrationSettings('Existing Entity Store', set),

        setQueryData: (value) => {
            set(
                produce((state: ExistingEntityState) => {
                    state.queryData = value;
                }),
                false,
                'Query Data Set'
            );
        },

        setConnectorName: (value) => {
            set(
                produce((state: ExistingEntityState) => {
                    state.connectorName = value;
                }),
                false,
                'Connector Name Set'
            );
        },

        setCreateNewTask: (value) => {
            set(
                produce((state: ExistingEntityState) => {
                    state.createNewTask = value;
                }),
                false,
                'Create New Task Flag Set'
            );
        },

        hydrateState: async (entityType, connectorId) => {
            if (connectorId) {
                const { setCreateNewTask, setHydrationErrorsExist } = get();

                const { data, error } = await getLiveSpecsByConnectorId(
                    entityType,
                    connectorId,
                    ['ops', 'demo']
                );

                if (error) {
                    setHydrationErrorsExist(true);
                    setCreateNewTask(true);
                } else if (data && data.length > 0) {
                    const { setQueryData, setConnectorName } = get();

                    setQueryData(data);
                    setConnectorName(data[0].title);
                } else {
                    setCreateNewTask(true);
                }
            }
        },

        resetState: () => {
            set(
                () => ({
                    ...getInitialStateData(),
                    ...getInitialHydrationData(),
                }),
                false,
                'Existing Entity State Reset'
            );
        },
    };
};

export const createExistingEntityStore = (key: ExistingEntityStoreNames) => {
    return create<ExistingEntityState>()(
        devtools((set, get) => getInitialState(set, get), devtoolsOptions(key))
    );
};
