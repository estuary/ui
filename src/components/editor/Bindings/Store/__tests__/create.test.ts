import { useBindingsEditorStore } from '../create';

describe('useBindingsEditorStore', () => {
    test('store need includes array with one element', () => {
        const testStore = useBindingsEditorStore.getState();
        expect(testStore).toMatchSnapshot();
    });
});
