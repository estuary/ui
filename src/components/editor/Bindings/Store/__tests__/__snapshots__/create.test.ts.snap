// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useBindingsEditorStore > store need includes array with one element 1`] = `
{
  "applyInferredSchema": [Function],
  "collectionData": null,
  "collectionInitializationAlert": null,
  "documentsRead": null,
  "editModeEnabled": false,
  "fullSourceConfigs": {},
  "fullSourceErrorsExist": false,
  "hasIncompatibleCollections": false,
  "incompatibleCollections": [],
  "inferSchemaResponse": null,
  "inferSchemaResponseDoneProcessing": false,
  "inferSchemaResponseEmpty": false,
  "inferSchemaResponseError": null,
  "inferSchemaResponse_Keys": [],
  "inferredSchemaApplicationErrored": false,
  "inferredSpec": null,
  "initializeSelections": [Function],
  "loadingInferredSchema": false,
  "populateInferSchemaResponse": [Function],
  "prefillFullSourceConfigs": [Function],
  "recommendFields": true,
  "removeFullSourceConfig": [Function],
  "resetState": [Function],
  "schemaInferenceDisabled": false,
  "schemaUpdateErrored": false,
  "schemaUpdated": true,
  "selectionSaving": false,
  "selections": {},
  "setCollectionData": [Function],
  "setCollectionInitializationAlert": [Function],
  "setDocumentsRead": [Function],
  "setEditModeEnabled": [Function],
  "setIncompatibleCollections": [Function],
  "setInferredSchemaApplicationErrored": [Function],
  "setInferredSpec": [Function],
  "setLoadingInferredSchema": [Function],
  "setRecommendFields": [Function],
  "setSchemaInferenceDisabled": [Function],
  "setSchemaUpdateErrored": [Function],
  "setSchemaUpdated": [Function],
  "setSelectionSaving": [Function],
  "setSingleSelection": [Function],
  "updateFullSourceConfig": [Function],
  "updateSchema": [Function],
}
`;
